### PR TITLE
Add intrinsic for getting type layout

### PR DESCRIPTION
**Pull request by jacob-hughes**
This PR adds the `core::gc::gc_layout<T>` intrinsic. It returns a bitmap of potential pointer locations for a type's layout. This is explained in more detail in each commit message.

### CODE REVIEW

**Comment by vext01** [comment]
Having read the first commit message, some comments:

> The core::mem::gc_layout::<T>() returns a pair of u64s: the first value is a bitmap representing which words in the type's layout are pointers; the second value is the size of the bitmap, (e.g. a size of 3 indicates only the last 3 bits in the bitmap are of interest).

The second element of the pair could be a u8, since it only needs to encode lengths of up to 64?

> Structs which have an alignment smaller than a machine word can be completely ignored by the collector, since they could never contain fields large enough to hold pointers.

Storing a pointer as (e.g.) two u32s is illegal, as that's pointer hiding, right?

File: compiler/rustc_mir/src/interpret/intrinsics/gc_layout.rs [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        // represent actual pointers would be purely accidental. Such
+        // structs should be ignored completely by the collector.
+        return vec![0; 16]; // empty (bitmap, size) (u64,u64) as byte stream
+    }
+
+    let src_field_iter = ty.ty_adt_def().unwrap().all_fields();
+    let mut bitmap: u64 = 0xFFFFFFFF;
+
+    for i in layout.fields.index_by_increasing_offset() {
+        let target_offset = layout.fields.offset(i as usize);
```
Comment (vext01) [inline compiler/rustc_mir/src/interpret/intrinsics/gc_layout.rs:30]: Is the `as` safe?

File: compiler/rustc_mir/src/interpret/intrinsics/gc_layout.rs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        // By default, the collector will conservatively scan a struct's
+        // memory, treating it as word-sized chunks. However, since rust is
+        // free to shuffle field order around, any in-memory words which
+        // represent actual pointers would be purely accidental. Such
+        // structs should be ignored completely by the collector.
+        return vec![0; 16]; // empty (bitmap, size) (u64,u64) as byte stream
+    }
+
+    let src_field_iter = ty.ty_adt_def().unwrap().all_fields();
+    let mut bitmap: u64 = 0xFFFFFFFF;
```
Comment (vext01) [inline compiler/rustc_mir/src/interpret/intrinsics/gc_layout.rs:27]: Do you need twice as many `F`s here? That's only a u32 constant. Could use `u64::MAX`?

File: compiler/rustc_mir/src/interpret/intrinsics/gc_layout.rs [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+        let layout_cx = LayoutCx { tcx, param_env };
+        let field_size = layout.field(&layout_cx, i).unwrap().size;
+
+        if field_size < w_size {
+            // If a field is smaller than a word, it will either be shuffled
+            // around in the struct or padded to the min alignment with
+            // uninitialized memory. Whatever the case, the word it lives in is
+            // never going to be a valid pointer.
+            let mask = target_offset.bytes() / w_align.bytes();
```
Comment (vext01) [inline compiler/rustc_mir/src/interpret/intrinsics/gc_layout.rs:40]: I'm not sure how that makes the correct mask. Does it need commentary?

File: src/test/ui/layout/gc-layout.rs [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    b: usize,
+    c: usize,
+}
+
+struct BigAlignNoTrace {
+    b: usize,
+    c: usize,
+    #[no_trace]
+    a: MultiFields,
+}
```
Comment (vext01) [inline src/test/ui/layout/gc-layout.rs:68]: Maybe add some edge cases? The empty struct? A `#[repr(C)]` struct? Do enums work?

File: library/core/src/gc.rs [hunk 5]
```diff
@@ -21,3 +21,18 @@ pub fn gc_layout<T>() -> (u64, u64) {
     let layout = intrinsics::gc_layout::<T>();
     (layout[0], layout[1])
 }
+
+#[unstable(feature = "gc", issue = "none")]
+#[cfg_attr(not(bootstrap), lang = "notrace")]
+pub auto trait NoTrace {}
+
+impl !NoTrace for usize {}
```
Comment (vext01) [inline library/core/src/gc.rs:29]: This is cool!

File: library/core/src/gc.rs [hunk 6]
```diff
@@ -21,3 +21,18 @@ pub fn gc_layout<T>() -> (u64, u64) {
...
+
+#[cfg(target_pointer_width = "64")]
+impl !NoTrace for u64 {}
+
+#[cfg(target_pointer_width = "32")]
+impl !NoTrace for u32 {}
+
+impl<T: ?Sized> !NoTrace for *mut T {}
+impl<T: ?Sized> !NoTrace for *const T {}
```
Comment (vext01) [inline library/core/src/gc.rs:38]: Do we have to do anything for smart pointer types?

**Comment by vext01** [comment]
A handful of minor comments.

**Comment by vext01** [comment]
Btw, you might consider using the bitfield crate to manage your bitmap?

**Comment by ltratt** [comment]
This looks good. Two main questions:

1. I think this can only cope with structs that contain *n* machine words where *n* is the bitsize of the machine. What happens if a struct is bigger than that? Do we fallback to conservative scanning?

2. What effect does this have on performance? e.g. if we (roughly) benchmark yksom before/after?

File: compiler/rustc_mir/src/interpret/intrinsics/gc_layout.rs (hunk 2 above)
Comment (jacob-hughes) [reply compiler/rustc_mir/src/interpret/intrinsics/gc_layout.rs:27]: You're right, `u64::MAX` would be better.

File: compiler/rustc_mir/src/interpret/intrinsics/gc_layout.rs (hunk 3 above)
Comment (jacob-hughes) [reply compiler/rustc_mir/src/interpret/intrinsics/gc_layout.rs:40]: `target_offset` is the byte in the struct's layout where this field begins. If we divide it by `w_align` (8 bytes) we get the offset in words. We then mask out the bit in the bitmap that corresponds to that fields word offset.

File: src/test/ui/layout/gc-layout.rs (hunk 4 above)
Comment (jacob-hughes) [reply src/test/ui/layout/gc-layout.rs:68]: No enums are not supported yet. Sure, I can add a couple of edge cases.

File: library/core/src/gc.rs (hunk 6 above)
Comment (jacob-hughes) [reply library/core/src/gc.rs:38]: We'll want to negatively implement `Gc<T>` but this isn't defined in `rustc_boehm` so that will happen in a separate repo.

Comment (ltratt) [reply library/core/src/gc.rs:38]: I guess Edd's also thinking of `Rc` and the like?

**Comment by jacob-hughes** [comment]
> Btw, you might consider using the bitfield crate to manage your bitmap?

@vext01 that would have certainly made things a bit easier, but I didn't want to add any dependencies to the const-eval phase of rustc.

**Comment by jacob-hughes** [comment]
> I think this can only cope with structs that contain n machine words where n is the bitsize of the machine. What happens if a struct is bigger than that? Do we fallback to conservative scanning?

Yes, at the moment we would fall back to conservative scanning. This is a limitation of Boehm, the collector interface only allows a maximum of a single machine word for passing type info (which can be repeated for collections). I think the reason for this is one of their tricks is that you can store this as a vtable field - something we don't have the luxury of in Rust. 

> What effect does this have on performance? e.g. if we (roughly) benchmark yksom before/after?

I don't have numbers yet. I'm working on a PR which actually makes use of this intrinsic in rboehm before it calls the allocator. That will let us know the impact of this, and I'll attach results as part of that.

File: library/core/src/gc.rs (hunk 6 above)
Comment (jacob-hughes) [reply library/core/src/gc.rs:38]: Ah I see. No I don't think so, `Rc` and the like will be _poisoned_ automatically if its contents is `!NoTrace`.  What we really care about here are raw pointers.

File: compiler/rustc_mir/src/interpret/intrinsics/gc_layout.rs (hunk 1 above)
Comment (jacob-hughes) [reply compiler/rustc_mir/src/interpret/intrinsics/gc_layout.rs:30]: Yes, the iterator will return nothing bigger than `u32`.

File: src/test/ui/layout/gc-layout.rs (hunk 4 above)
Comment (jacob-hughes) [reply src/test/ui/layout/gc-layout.rs:68]: Here is the ZST case: eb9cd6f706200a93072e904c1ba988df365ac0b6

File: compiler/rustc_mir/src/interpret/intrinsics/gc_layout.rs (hunk 2 above)
Comment (jacob-hughes) [reply compiler/rustc_mir/src/interpret/intrinsics/gc_layout.rs]: Fixed: dfa5cafce3c376f804980b41253f030f2d928bc6

File: compiler/rustc_mir/src/interpret/intrinsics/gc_layout.rs (hunk 1 above)
Comment (ltratt) [reply compiler/rustc_mir/src/interpret/intrinsics/gc_layout.rs:30]: I think we've gradually come to the conclusion that we want to avoid `as` if there's another mechanism. `usize::from(i)` if possible or `usize::try_from(i).unwrap()` otherwise (we assume the compiler removes the costs of the latter if it's infallible).

**Comment by ltratt** [comment]
> Yes, at the moment we would fall back to conservative scanning.

Just to be clear: "would" or "do"? [I'm fine with the limitation, but I want to make sure we really do fall back to conservative scanning! We should also add a test case that checks this.]

File: compiler/rustc_mir/src/interpret/intrinsics/gc_layout.rs (hunk 3 above)
Comment (vext01) [reply compiler/rustc_mir/src/interpret/intrinsics/gc_layout.rs:40]: Does that rather good explanation need to be a comment?

File: compiler/rustc_mir/src/interpret/intrinsics/gc_layout.rs (hunk 1 above)
Comment (jacob-hughes) [reply compiler/rustc_mir/src/interpret/intrinsics/gc_layout.rs:30]: While I agree in principle, rustc is littered with `as` casting of this kind. In fact this code was lifted from an identical pattern in the llvm codegen phase. I could buy this argument for our own crates, but here it feels like shutting the stable door after the horse has bolted.

Comment (ltratt) [reply compiler/rustc_mir/src/interpret/intrinsics/gc_layout.rs:30]: That's because most of the code in rustc predates `from`/`try_from`. I don't think there's a good reason for us to perpetuate that mistake, especially because `as` is so dangerous.

**Comment by jacob-hughes** [comment]
> Just to be clear: "would" or "do"? [I'm fine with the limitation, but I want to make sure we really do fall back to conservative scanning! We should also add a test case that checks this.]

Well it's up to the implementer (rboehm in this case) checks the type's size will fit before passing it to `gc_layout`. I guess we could add an assertion?

File: compiler/rustc_mir/src/interpret/intrinsics/gc_layout.rs (hunk 1 above)
Comment (jacob-hughes) [reply compiler/rustc_mir/src/interpret/intrinsics/gc_layout.rs]: Ok, fixed here 6b5b7565856b618372289638d57424354e71f288

File: compiler/rustc_mir/src/interpret/intrinsics/gc_layout.rs (hunk 3 above)
Comment (jacob-hughes) [reply compiler/rustc_mir/src/interpret/intrinsics/gc_layout.rs:43]: Sure, how about 43d49a3836e3f588da0e25e777e7b9a6240944c7?

**Comment by vext01** [comment]
LGTM.

**Comment by ltratt** [comment]
> Well it's up to the implementer (rboehm in this case) to check the type's size will fit before passing it to gc_layout. I guess we could add an assertion?

I think we need to do *something*. If it's something the caller has to respect then I think we should document that it's a `panic`able scenario.

**Comment by jacob-hughes** [comment]
How about something like this d692e09?

**Comment by jacob-hughes** [comment]
> If you think the branch would be too expensive

This is my main concern. It will be part of the hot path in allocation. I'd prefer to make this function unsafe.

I see no harm in keeping the debug assertion too, what do you think?

**Comment by ltratt** [comment]
If it's explicitly marked `unsafe` then a `debug_assert` is fine.

**Comment by jacob-hughes** [comment]
Try this: e2ecf538b4058123205bb24d11aa65724dd705cc

**Comment by ltratt** [comment]
OK, let's get the benchmark numbers here, but once we've seen those, I expect we'll be ready for squashing.

**Comment by jacob-hughes** [comment]
Here are some rough benchmark numbers of a first run with precise type maps on yksom:
```
  Benchmark     Executor   Suite   Extra   Core   Size   Var   #Samples   Mean (ms)
  BubbleSort    baseline   micro       3      1                      10   66
  BubbleSort    precise    micro       3      1                      10   80
  Dispatch      baseline   micro       2      1                      10   35
  Dispatch      precise    micro       2      1                      10   55
  Fannkuch      baseline   micro       6      1                      10   30
  Fannkuch      precise    micro       6      1                      10   40
  Fibonacci     baseline   micro       3      1                      10   128
  Fibonacci     precise    micro       3      1                      10   190
  FieldLoop     baseline   micro       1      1                      10   29
  FieldLoop     precise    micro       1      1                      10   39
  IntegerLoop   baseline   micro       2      1                      10   68
  IntegerLoop   precise    micro       2      1                      10   103
  JsonSmall     baseline   macro       1      1                      10   59
  JsonSmall     precise    macro       1      1                      10   76
  List          baseline   micro       2      1                      10   68
  List          precise    micro       2      1                      10   86
  Loop          baseline   micro       5      1                      10   87
  Loop          precise    micro       5      1                      10   127
  Mandelbrot    baseline   micro      30      1                      10   165
  Mandelbrot    precise    micro      30      1                      10   165
  Permute       baseline   micro       3      1                      10   52
  Permute       precise    micro       3      1                      10   67
  Queens        baseline   micro       2      1                      10   46
  Queens        precise    micro       2      1                      10   59
  QuickSort     baseline   micro       1      1                      10   19
  QuickSort     precise    micro       1      1                      10   28
  Recurse       baseline   micro       3      1                      10   64
  Recurse       precise    micro       3      1                      10   87
  Richards      baseline   macro       1      1                      10   934
  Richards      precise    macro       1      1                      10   1325
  Sieve         baseline   micro       4      1                      10   80
  Sieve         precise    micro       4      1                      10   115
  Storage       baseline   micro       1      1                      10   22
  Storage       precise    micro       1      1                      10   29
  Sum           baseline   micro       2      1                      10   35
  Sum           precise    micro       2      1                      10   55
  Towers        baseline   micro       2      1                      10   55
  Towers        precise    micro       2      1                      10   67
  TreeSort      baseline   micro       1      1                      10   54
  TreeSort      precise    micro       1      1                      10   77
  WhileLoop     baseline   micro      10      1                      10   76
  WhileLoop     precise    micro      10      1                      10   104
```

Right now they don't look promising, but I don't think they tell us much about this PR - this intrinsic is a constexpr, and will have very little impact in runtime cost. The problem is that I implemented precise marking very roughly in yksom, and the results here tell us more about how we need to fix that code rather than this PR.

**Comment by ltratt** [comment]
Maybe you can make a public yksom branch so I can see what changes you've made there? It's a bit difficult to guess what they might be!

**Comment by jacob-hughes** [comment]
It's a little tricky because there are natrob and rboehm changes too. I can give it a try though if you might find it useful.

**Comment by ltratt** [comment]
I might have been a bit hasty: really, it's up to you. If don't need comments quite yet, public branches might not be necessary.

**Comment by jacob-hughes** [comment]
Let's leave it if that's ok. I'm going to post these as PRs and I think it will help if I give you some context in the PR descriptions.

**Comment by ltratt** [comment]
Works for me!

**Comment by jacob-hughes** [comment]
Can I squash ? :)

**Comment by ltratt** [comment]
I think we need to see what we can do to improve the numbers before we integrate this. There is a chance -- a small chance, I think, but not quite zero -- that this might always make performance worse. If that was the case, then I don't think we'd want to integrate it?

**Comment by jacob-hughes** [comment]
I don't mind waiting, however, this PR is just the first step and doesn't really have any bearing on performance. It's orthogonal really because it's just a function in the compiler that lets us get the type layout. 

It may be the case that precise object marking doesn't give us the performance boost we hope for in yksom, but that doesn't mean this isn't going to be a useful function for other users.

**Comment by ltratt** [comment]
To make sure that I understand: if someone doesn't explicitly opt-in to the features this PR enables, there will be no change in the performance they see?

**Comment by jacob-hughes** [comment]
Yes that's correct. This feature just adds `core::gc::gc_layout::<T>()` to libcore.

**Comment by ltratt** [comment]
Got it. Please squash.

**Comment by jacob-hughes** [comment]
This kept failing the main test suite for me locally. It seems to be a problem on master too, so maybe I've inherited a bug from upstream. 

bors try

**Comment by ltratt** [comment]
What failure are you seeing?

**Comment by jacob-hughes** [comment]
Seeing failures on the big integration tests which poke around at the gdb output.

```
---- [debuginfo-gdb] debuginfo/pretty-huge-vec.rs stdout ----
NOTE: compiletest thinks it is using GDB with native rust support
NOTE: compiletest thinks it is using GDB version 10001000

error: line not found in debugger output: $2 = &[u8] {data_ptr: [...]"\000", length: 1000000000}
status: exit code: 0
command: "/usr/bin/gdb" "-quiet" "-batch" "-nx" "-command=/home/jake/research/rustc_boehm/build/x86_64-unknown-linux-gnu/test/debuginfo/pretty-huge-vec.gdb/pretty-huge-vec.debugger.script"
stdout:
------------------------------------------
GNU gdb (GDB) 10.1
Copyright (C) 2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-pc-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word".
Breakpoint 1 at 0x1fc5: file /home/jake/research/rustc_boehm/src/test/debuginfo/pretty-huge-vec.rs, line 27.
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib/libthread_db.so.1".

Breakpoint 1, pretty_huge_vec::main () at /home/jake/research/rustc_boehm/src/test/debuginfo/pretty-huge-vec.rs:27
27          zzz(); // #break
$1 = Vec(size=1000000000) = {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0...}
$2 = &[u8] {data_ptr: 0x7fffbb91f010, length: 1000000000}
A debugging session is active.

        Inferior 1 [process 791010] will be killed.

Quit anyway? (y or n) [answered Y; input not from terminal]

------------------------------------------
stderr:
------------------------------------------

------------------------------------------



failures:
    [debuginfo-gdb] debuginfo/extern-c-fn.rs
    [debuginfo-gdb] debuginfo/generator-objects.rs
    [debuginfo-gdb] debuginfo/issue-57822.rs
    [debuginfo-gdb] debuginfo/pretty-huge-vec.rs

test result: FAILED. 
```

**Comment by ltratt** [comment]
That I have no idea about :)

**Comment by vext01** [comment]
Is this a gdb version mismatch?

It expected to match:
```
$2 = &[u8] {data_ptr: [...]"\000", length: 1000000000}
```

but saw:
```
$2 = &[u8] {data_ptr: 0x7fffbb91f010, length: 1000000000}
```

Notice how the `"\000"` is absent in the actual output...

Where is that gdb 10 coming from? In-tree?

**Comment by jacob-hughes** [comment]
bors try

**Comment by jacob-hughes** [comment]
bors try

**Comment by jacob-hughes** [comment]
bors try

**Comment by jacob-hughes** [comment]
bors try

**Comment by jacob-hughes** [comment]
bors try

**Comment by jacob-hughes** [comment]
Finally. After faffing around with the linkchecker and `--no-doc` (which is broken on nightly!), this is ready for squashing again.

**Comment by ltratt** [comment]
Please squash.

**Comment by jacob-hughes** [comment]
Squashed

**Comment by ltratt** [comment]
Can we document in the appropriate commits that we disable some doc tests? Feel free to force push an

**Comment by jacob-hughes** [comment]
Sorry, I should have been more clear: I haven't disabled them, I ended up fixing the problems that `NoTrace` caused as it was easier. (Disabling the doc tests is broken on nightly at the moment)

**Comment by ltratt** [comment]
That's even better 👍:)

bors r+
